### PR TITLE
Add skeletons for libpod container, pod, state

### PR DESCRIPTION
**Pull request by mheon**
Code was starting to get a little big, so I'd like to get this in before I drop the actual in-memory state implementation in. Adds the basic skeletons for containers and pods, plus an interface for the runtime state.

### CODE REVIEW

**Comment by mheon** [comment]
@rhatdan PTAL

File: libpod/container.go [hunk 1]
```diff
@@ -1,69 +1,126 @@
...
+	spec := new(spec.Spec)
+	deepcopier.Copy(c.spec).To(spec)
+
+	return spec
+}
+
+// Make a new container
+func newContainer(rspec *spec.Spec) (*Container, error) {
+	if rspec == nil {
+		return nil, fmt.Errorf("must provide a valid spec to construct container")
```
Comment (rhatdan) [inline libpod/container.go]: We should either wrap errors with syscall.EINVAL, or define out own errors.

File: libpod/container.go [hunk 2]
```diff
@@ -1,69 +1,126 @@
...
+	}
+
+	ctr := new(Container)
+	ctr.id = stringid.GenerateNonCryptoID()
+	ctr.name = ctr.id // TODO generate unique human-readable names
```
Comment (rhatdan) [inline libpod/container.go]: Should we steal the moby code for this?

File: libpod/pod.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+// ID retrieves the pod's ID
+func (p *Pod) ID() string {
+	return p.id
+}
+
+// Name retrieves the pod's name
+func (p *Pod) Name() string {
+	return p.name
```
Comment (rhatdan) [inline libpod/pod.go]: Are pod names changeable, Do we need locking here?

File: libpod/pod.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+// Adds a container to the pod
+// Does not check that container's pod ID is set correctly, or attempt to set
+// pod ID after adding
+func (p *Pod) addContainer(ctr *Container) error {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	if !p.valid {
+		return fmt.Errorf("pod has already been removed")
```
Comment (rhatdan) [inline libpod/pod.go]: Lets make error constants.

File: libpod/runtime.go [hunk 5]
```diff
@@ -136,81 +137,244 @@ func (r *Runtime) Shutdown(force bool) error {
...
 
 // NewContainer creates a new container from a given OCI config
-func (r *Runtime) NewContainer(spec *spec.Spec, options ...CtrCreateOption) (*ctr.Container, error) {
-	return nil, ctr.ErrNotImplemented
+func (r *Runtime) NewContainer(spec *spec.Spec, options ...CtrCreateOption) (*Container, error) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	if !r.valid {
+		return nil, fmt.Errorf("runtime has already been shut down")
```
Comment (rhatdan) [inline libpod/runtime.go]: You should create these with errors and set up constants.

File: libpod/runtime.go [hunk 6]
```diff
@@ -136,81 +137,244 @@ func (r *Runtime) Shutdown(force bool) error {
...
+	return r.state.LookupContainer(idOrName)
 }
 
 // GetContainers retrieves all containers from the state
 // Filters can be provided which will determine what containers are included in
 // the output. Multiple filters are handled by ANDing their output, so only
 // containers matching all filters are returned
-func (r *Runtime) GetContainers(filters ...ContainerFilter) ([]*ctr.Container, error) {
-	return nil, ctr.ErrNotImplemented
+func (r *Runtime) GetContainers(filters ...ContainerFilter) ([]*Container, error) {
```
Comment (rhatdan) [inline libpod/runtime.go]: Should this just be Containers? Drop the Get?

File: libpod/runtime.go [hunk 7]
```diff
@@ -136,81 +137,244 @@ func (r *Runtime) Shutdown(force bool) error {
...
+	return r.state.LookupPod(idOrName)
 }
 
 // GetPods retrieves all pods
 // Filters can be provided which will determine which pods are included in the
 // output. Multiple filters are handled by ANDing their output, so only pods
 // matching all filters are returned
-func (r *Runtime) GetPods(filters ...PodFilter) ([]*pod.Pod, error) {
-	return nil, ctr.ErrNotImplemented
+func (r *Runtime) GetPods(filters ...PodFilter) ([]*Pod, error) {
```
Comment (rhatdan) [inline libpod/runtime.go]: Should this just be Pods?  Drop the Get?

**Comment by rhatdan** [comment]
@mrunalp @umohnani8 @mitrmac PTAL

File: libpod/container.go (hunk 2 above)
Comment (mheon) [reply libpod/container.go]: I'd be fine with this, but this PR is already huge, so I'd prefer to leave the extra vendoring to later.

File: libpod/pod.go (hunk 3 above)
Comment (mheon) [reply libpod/pod.go]: I'd prefer to not allow pods or containers be renamed, but I'm not strongly opposed to adding this

File: libpod/runtime.go (hunk 6 above)
Comment (mheon) [reply libpod/runtime.go]: We had GetContainer, so I made this match. I think Containers might be better, though, it's less easy to confuse the two that way.

File: libpod/container.go (hunk 2 above)
Comment (rhatdan) [reply libpod/container.go]: SGTM

File: libpod/pod.go (hunk 3 above)
Comment (rhatdan) [reply libpod/pod.go]: I would prefer that they not be renamed, but I don't know what is expected.  @mrunalp WDYT

**Comment by mheon** [comment]
Comments addressed. @rhatdan @mrunalp can we get this reviewed and merged? Got a few other patches depending on this.

**Comment by rhatdan** [comment]
Travis failures.

**Comment by mheon** [comment]
Gofmt issue, fixed it

File: libpod/runtime.go [hunk 8]
```diff
@@ -1,13 +1,10 @@
 package libpod
```
Comment (rhatdan) [inline libpod/runtime.go]: Should we break runtime.go into multiple files, container.go, image,go, pod.go  These will get rather large once we port all of likpod into them.

**Comment by rhatdan** [comment]
Overall this LGTM, I just think we should break up into multiple go files. But I am fine with doing this in another PR.

File: libpod/runtime.go (hunk 8 above)
Comment (mheon) [reply libpod/runtime.go]: Yeah, that sounds like a plan. It's a trivial change, I think we can make it in here.

Comment (mheon) [reply libpod/runtime.go]: Fixed

**Comment by mheon** [comment]
Travis passed, starting jenkins

**Comment by mheon** [comment]
[retest]

**Comment by rhatdan** [comment]
I think we need to get @mrunalp To comment and then we could merge?

**Comment by mheon** [comment]
Agreed, would be nice to have another set of eyes on the review.

**Comment by mheon** [comment]
Rebased so Travis will rerun. Should not affect any of the E2E tests. @mrunalp PTAL

**Comment by rhatdan** [comment]
[retest]

**Comment by rhatdan** [comment]
@mrunalp Tests pass.  PTAL

**Comment by mheon** [comment]
@rhatdan Rebased, let's get the integration tests running once travis is green

**Comment by rhatdan** [comment]
@cevich tests did not run?  Any ideas?

**Comment by mheon** [comment]
Rebased and force-pushed again. @rhatdan try the tests again?

**Comment by mheon** [comment]
@mrunalp This looks like just the guestbook failure, which I think is expected as of now? Can you take a quick look and make sure this is good to merge?

**Comment by mheon** [comment]
Rebased after conflicts from another PR, going to retrigger the tests

**Comment by rhatdan** [comment]
148/149 on E2E tests pass, with expected failure.  Integration test passes.  Merging.

**Pull request merged**